src/JavascriptVM/vm.ts

```diff
@@ -26,20 +48,110 @@ export class BlocklyInterpreter {
       interpreter.setProperty(globals, "alert", alert);
       interpreter.setProperty(globals, "highlightBlock", highlightBlock);
     });
+
+    // Start running in a paused state - ie, spawn a "thread"
+    this._run();
   }
 
-  step(): boolean {
+  /**
+   * Executes a "block" of work. If there is no more work to be executed, stops execution and
+   * fires the "onFinished" callback if present. This is the main function that should be called
+   * to execute code on the VM.
+   */
+  private _step(): boolean {
     let finished = false;
-    while (!this.pause && !finished) {
+    this.blockHighlighted = false;
+    while (!this.blockHighlighted && !finished) {
       finished = !this.interpreter.step();
     }
+    this.blockHighlighted = false;
+
+    if (finished) {
+      // put the VM into a "stopped" state
+      this.stop();
 
-    this.pause = false;
+      // alert the client of the VM that execution is finished if there is a cb registered
+      this.callbacks.onFinish && this.callbacks.onFinish();
+    }
 
     return finished;
   }
 
-  run(): boolean {
-    return this.interpreter.run();
+  /**
+   * Steps execution by one block if there is code left to execute. Returns true if there is more to execute, false otherwise.
+   */
+  step(): boolean {
+    // do not step an already running program
+    if (this.executionState === ExecutionState.RUNNING) {
+      return true;
+    }
+
+    // if stopped, do nothing.
+    if (this.executionState === ExecutionState.STOPPED) {
+      return false;
+    }
+
+    return this._step();
+  }
+
+  /**
+   * Continually steps the program at a cadence until there is nothing left to run or until the user stops the VM.
+   */
+  private _run() {
+    // do not schedule any more work if stopped
+    if (this.executionState === ExecutionState.STOPPED) {
+      return;
+    }
+
+    setTimeout(() => {
+      // if paused - schedule the next frame
+      if (this.executionState === ExecutionState.PAUSED) {
+        return this._run();
+      }
+
+      this._step();
+
+      // schedule the next run
+      return this._run();
+    }, 1000 / BLOCKS_PER_SECOND);
```

Comment:
I think it'd be cool if we also exposed this in some way to the user so they could set the execution "speed" of their program which is why I went with setTimeout and not `requestAnimationFrame` or `setInterval`. Maybe a turtle / hare option? Good follow up issue, nonetheless.